CSR layout: CPU  addmm Fix

aocsa opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57403 CUDA support in the CSR layout: matvec
* **#58622 CSR layout: CPU  addmm Fix**
* #57275 CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr
* #57274 CUDA support in the CSR layout: constructors
 
Fixes gh-58757


I added this PR before CUDA addmm/matvec #57403 with a fix for CPU version of addmm op. Just for context, before this PR, only `CSR @ vector` was supported. I found out a minor bug in the `addmm_out_sparse_csr_dense_cpu` for the non MKL code which is solved in this PR. 

Moreover, I discovered a limitation in the current MKL implementation. It only works well (acceptable tolerance for output error) with square matrices. I was looking in deep to this issue and I found out that it could be a limitation of the MKL API.

I used this [gist code](https://gist.github.com/aocsa/0606e833cd16a8bfb7d37a5fbb3a5b14) based on [this](https://github.com/baidu-research/DeepBench/blob/master/code/intel/spmm/spmm_bench.cpp) to test this behavior. 

As you can see there is not an acceptable output error (last column) when the matrices are squares and there is a not acceptable error when the matrices are not square:

```
             m   |   n  |   k  |    nnz   | time spmv  | perf spmv  |  error
sparse    [    10|    10|    10|         3| 2.83199e-06|   0.0197315|           0
sparse    [     4|     4|     4|         3| 2.77099e-06|  0.00806635|           0
sparse    [     6|     6|     6|         1| 2.79799e-06|  0.00399425|           0
sparse    [    20|    20|    20|        30| 5.80697e-06|    0.192456|           0
sparse    [   200|   200|   200|      2006| 5.28837e-05|     14.1309|   1.493e-06
sparse    [  2000|  2000|  2000|    199850|   0.0341477|     21.8023|  0.00919917
sparse    [    20|     1|    10|        12| 8.04476e-05| 0.000277842|           0
sparse    [   107|     1|   384|      2084| 3.14868e-05|    0.123282|           0
sparse    [  1008|     1|   184|      9350| 2.15189e-05|    0.809323|           0
sparse    [ 10752|     1|  3584|   1925372| 0.000159745|       22.45|    0.217587
sparse    [  7860|  6000|  2560|   1006470|    0.520181|     21.6236|   0.0550456
sparse    [ 10752|     1|  3584|   1927337| 0.000151101|     23.7585|    0.217443
sparse    [ 10752|     2|  3584|   1928639|  0.00027718|     25.9209|    0.275867
sparse    [ 10752|     4|  3584|   1927072| 0.000824986|     17.4037|    0.278116
sparse    [ 10752|  1500|  3584|   1926182|    0.181975|     29.5737|    0.251635
sparse    [ 10752|  3000|  3584|   1927537|    0.319327|     33.7302|    0.258285
sparse    [ 10752|  6000|  3584|   1925716|    0.648246|     33.1997|    0.226093
sparse    [  7680|     1|  2560|   1966992| 0.000147336|      24.867|    0.369858 
... 

```

Looking forward to your comments @v0dro, @pearu.

aocsa committed:
CSR layout: CPU  addmm Fix

[ghstack-poisoned]

aocsa committed:
addmm cuda on "CSR layout: CPU  addmm Fix"




[ghstack-poisoned]

aocsa committed:
rebased/addressed review comments - 3 on "CSR layout: CPU  addmm Fix"




[ghstack-poisoned]

aocsa committed:
rebased/last feedback addressed on "CSR layout: CPU  addmm Fix"

 
I added this PR before CUDA addmm/matvec #57403 with a fix for CPU version of addmm op. Just for context, before this PR, only `CSR @ vector` was supported. I found out a minor bug in the `addmm_out_sparse_csr_dense_cpu` for the non MKL code which is solved in this PR. 

Moreover, I discovered a limitation in the current MKL implementation. It only works with square matrices. I was looking in deep to this issue and I found out that it could be a limitation of the MKL API. 

I used this [gist code](https://gist.github.com/aocsa/0606e833cd16a8bfb7d37a5fbb3a5b14) based on [this](https://github.com/baidu-research/DeepBench/blob/master/code/intel/spmm/spmm_bench.cpp) to test this behavior. 

As you can see there is not an error (last column) when the matrices are squares and there is a little error when the matrices are not square:

```
             m   |   n  |   k  |    nnz   | time spmv  | perf spmv  |  error
sparse    [    10|    10|    10|         3| 2.37201e-06|   0.0235578|           0
sparse    [     4|     4|     4|         3| 1.79801e-06|   0.0124314|           0
sparse    [     6|     6|     6|         1| 1.76501e-06|  0.00633191|           0
sparse    [    20|    20|    20|        30| 5.12503e-06|    0.218065|           0
sparse    [    20|     1|    10|        12| 1.33501e-06|   0.0167428|           0
sparse    [   107|     1|   384|      2025| 1.39701e-06|     2.69996| 3.45707e-06
sparse    [  1008|     1|   184|      9179| 2.84901e-06|      6.0011|           0
sparse    [ 10752|     1|  3584|   1927625| 8.54424e-05|     42.0222|     0.21477
sparse    [  7860|  6000|  2560|   1005021|    0.284707|     39.4511|   0.0547853
sparse    [ 10752|     1|  3584|   1927289| 0.000149735|     23.9748|    0.215473
sparse    [ 10752|     2|  3584|   1927367| 0.000275535|     26.0584|     0.27695
sparse    [ 10752|     4|  3584|   1925245| 0.000812662|     17.6509|    0.277165

```

Looking forward to your comments @v0dro, @pearu. 






[ghstack-poisoned]

aocsa committed:
run CI/last feedback addressed on "CSR layout: CPU  addmm Fix"

 
I added this PR before CUDA addmm/matvec #57403 with a fix for CPU version of addmm op. Just for context, before this PR, only `CSR @ vector` was supported. I found out a minor bug in the `addmm_out_sparse_csr_dense_cpu` for the non MKL code which is solved in this PR. 

Moreover, I discovered a limitation in the current MKL implementation. It only works with square matrices. I was looking in deep to this issue and I found out that it could be a limitation of the MKL API. 

I used this [gist code](https://gist.github.com/aocsa/0606e833cd16a8bfb7d37a5fbb3a5b14) based on [this](https://github.com/baidu-research/DeepBench/blob/master/code/intel/spmm/spmm_bench.cpp) to test this behavior. 

As you can see there is not an error (last column) when the matrices are squares and there is a little error when the matrices are not square:

```
             m   |   n  |   k  |    nnz   | time spmv  | perf spmv  |  error
sparse    [    10|    10|    10|         3| 2.37201e-06|   0.0235578|           0
sparse    [     4|     4|     4|         3| 1.79801e-06|   0.0124314|           0
sparse    [     6|     6|     6|         1| 1.76501e-06|  0.00633191|           0
sparse    [    20|    20|    20|        30| 5.12503e-06|    0.218065|           0
sparse    [    20|     1|    10|        12| 1.33501e-06|   0.0167428|           0
sparse    [   107|     1|   384|      2025| 1.39701e-06|     2.69996| 3.45707e-06
sparse    [  1008|     1|   184|      9179| 2.84901e-06|      6.0011|           0
sparse    [ 10752|     1|  3584|   1927625| 8.54424e-05|     42.0222|     0.21477
sparse    [  7860|  6000|  2560|   1005021|    0.284707|     39.4511|   0.0547853
sparse    [ 10752|     1|  3584|   1927289| 0.000149735|     23.9748|    0.215473
sparse    [ 10752|     2|  3584|   1927367| 0.000275535|     26.0584|     0.27695
sparse    [ 10752|     4|  3584|   1925245| 0.000812662|     17.6509|    0.277165

```

Looking forward to your comments @v0dro, @pearu. 






[ghstack-poisoned]

aocsa committed:
rebased needed on "CSR layout: CPU  addmm Fix"

 
Fixes gh-58757


I added this PR before CUDA addmm/matvec #57403 with a fix for CPU version of addmm op. Just for context, before this PR, only `CSR @ vector` was supported. I found out a minor bug in the `addmm_out_sparse_csr_dense_cpu` for the non MKL code which is solved in this PR. 

Moreover, I discovered a limitation in the current MKL implementation. It only works with square matrices. I was looking in deep to this issue and I found out that it could be a limitation of the MKL API. 

I used this [gist code](https://gist.github.com/aocsa/0606e833cd16a8bfb7d37a5fbb3a5b14) based on [this](https://github.com/baidu-research/DeepBench/blob/master/code/intel/spmm/spmm_bench.cpp) to test this behavior. 

As you can see there is not an error (last column) when the matrices are squares and there is a little error when the matrices are not square:

```
             m   |   n  |   k  |    nnz   | time spmv  | perf spmv  |  error
sparse    [    10|    10|    10|         3| 2.83199e-06|   0.0197315|           0
sparse    [     4|     4|     4|         3| 2.77099e-06|  0.00806635|           0
sparse    [     6|     6|     6|         1| 2.79799e-06|  0.00399425|           0
sparse    [    20|    20|    20|        30| 5.80697e-06|    0.192456|           0
sparse    [   200|   200|   200|      2006| 5.28837e-05|     14.1309|   1.493e-06
sparse    [  2000|  2000|  2000|    199850|   0.0341477|     21.8023|  0.00919917
sparse    [    20|     1|    10|        12| 8.04476e-05| 0.000277842|           0
sparse    [   107|     1|   384|      2084| 3.14868e-05|    0.123282|           0
sparse    [  1008|     1|   184|      9350| 2.15189e-05|    0.809323|           0
sparse    [ 10752|     1|  3584|   1925372| 0.000159745|       22.45|    0.217587
sparse    [  7860|  6000|  2560|   1006470|    0.520181|     21.6236|   0.0550456
sparse    [ 10752|     1|  3584|   1927337| 0.000151101|     23.7585|    0.217443
sparse    [ 10752|     2|  3584|   1928639|  0.00027718|     25.9209|    0.275867
sparse    [ 10752|     4|  3584|   1927072| 0.000824986|     17.4037|    0.278116
sparse    [ 10752|  1500|  3584|   1926182|    0.181975|     29.5737|    0.251635
sparse    [ 10752|  3000|  3584|   1927537|    0.319327|     33.7302|    0.258285
sparse    [ 10752|  6000|  3584|   1925716|    0.648246|     33.1997|    0.226093
sparse    [  7680|     1|  2560|   1966992| 0.000147336|      24.867|    0.369858 
... 

```

Looking forward to your comments @v0dro, @pearu. 






[ghstack-poisoned]

aocsa committed:
rebased needed : on "CSR layout: CPU  addmm Fix"

 
Fixes gh-58757


I added this PR before CUDA addmm/matvec #57403 with a fix for CPU version of addmm op. Just for context, before this PR, only `CSR @ vector` was supported. I found out a minor bug in the `addmm_out_sparse_csr_dense_cpu` for the non MKL code which is solved in this PR. 

Moreover, I discovered a limitation in the current MKL implementation. It only works with square matrices. I was looking in deep to this issue and I found out that it could be a limitation of the MKL API. 

I used this [gist code](https://gist.github.com/aocsa/0606e833cd16a8bfb7d37a5fbb3a5b14) based on [this](https://github.com/baidu-research/DeepBench/blob/master/code/intel/spmm/spmm_bench.cpp) to test this behavior. 

As you can see there is not an error (last column) when the matrices are squares and there is a little error when the matrices are not square:

```
             m   |   n  |   k  |    nnz   | time spmv  | perf spmv  |  error
sparse    [    10|    10|    10|         3| 2.83199e-06|   0.0197315|           0
sparse    [     4|     4|     4|         3| 2.77099e-06|  0.00806635|           0
sparse    [     6|     6|     6|         1| 2.79799e-06|  0.00399425|           0
sparse    [    20|    20|    20|        30| 5.80697e-06|    0.192456|           0
sparse    [   200|   200|   200|      2006| 5.28837e-05|     14.1309|   1.493e-06
sparse    [  2000|  2000|  2000|    199850|   0.0341477|     21.8023|  0.00919917
sparse    [    20|     1|    10|        12| 8.04476e-05| 0.000277842|           0
sparse    [   107|     1|   384|      2084| 3.14868e-05|    0.123282|           0
sparse    [  1008|     1|   184|      9350| 2.15189e-05|    0.809323|           0
sparse    [ 10752|     1|  3584|   1925372| 0.000159745|       22.45|    0.217587
sparse    [  7860|  6000|  2560|   1006470|    0.520181|     21.6236|   0.0550456
sparse    [ 10752|     1|  3584|   1927337| 0.000151101|     23.7585|    0.217443
sparse    [ 10752|     2|  3584|   1928639|  0.00027718|     25.9209|    0.275867
sparse    [ 10752|     4|  3584|   1927072| 0.000824986|     17.4037|    0.278116
sparse    [ 10752|  1500|  3584|   1926182|    0.181975|     29.5737|    0.251635
sparse    [ 10752|  3000|  3584|   1927537|    0.319327|     33.7302|    0.258285
sparse    [ 10752|  6000|  3584|   1925716|    0.648246|     33.1997|    0.226093
sparse    [  7680|     1|  2560|   1966992| 0.000147336|      24.867|    0.369858 
... 

```

Looking forward to your comments @v0dro, @pearu. 






[ghstack-poisoned]

malfet reviewed (changes requested):
Can you please make this change independent from the rest of the stack?
And also, keep it restricted to PR title (i.e. just fixing addmm bug)

aocsa committed:
support for MKL matvec on "CSR layout: CPU  addmm Fix"

 
Fixes gh-58757


I added this PR before CUDA addmm/matvec #57403 with a fix for CPU version of addmm op. Just for context, before this PR, only `CSR @ vector` was supported. I found out a minor bug in the `addmm_out_sparse_csr_dense_cpu` for the non MKL code which is solved in this PR. 

Moreover, I discovered a limitation in the current MKL implementation. It only works with square matrices. I was looking in deep to this issue and I found out that it could be a limitation of the MKL API. 

I used this [gist code](https://gist.github.com/aocsa/0606e833cd16a8bfb7d37a5fbb3a5b14) based on [this](https://github.com/baidu-research/DeepBench/blob/master/code/intel/spmm/spmm_bench.cpp) to test this behavior. 

As you can see there is not an error (last column) when the matrices are squares and there is a little error when the matrices are not square:

```
             m   |   n  |   k  |    nnz   | time spmv  | perf spmv  |  error
sparse    [    10|    10|    10|         3| 2.83199e-06|   0.0197315|           0
sparse    [     4|     4|     4|         3| 2.77099e-06|  0.00806635|           0
sparse    [     6|     6|     6|         1| 2.79799e-06|  0.00399425|           0
sparse    [    20|    20|    20|        30| 5.80697e-06|    0.192456|           0
sparse    [   200|   200|   200|      2006| 5.28837e-05|     14.1309|   1.493e-06
sparse    [  2000|  2000|  2000|    199850|   0.0341477|     21.8023|  0.00919917
sparse    [    20|     1|    10|        12| 8.04476e-05| 0.000277842|           0
sparse    [   107|     1|   384|      2084| 3.14868e-05|    0.123282|           0
sparse    [  1008|     1|   184|      9350| 2.15189e-05|    0.809323|           0
sparse    [ 10752|     1|  3584|   1925372| 0.000159745|       22.45|    0.217587
sparse    [  7860|  6000|  2560|   1006470|    0.520181|     21.6236|   0.0550456
sparse    [ 10752|     1|  3584|   1927337| 0.000151101|     23.7585|    0.217443
sparse    [ 10752|     2|  3584|   1928639|  0.00027718|     25.9209|    0.275867
sparse    [ 10752|     4|  3584|   1927072| 0.000824986|     17.4037|    0.278116
sparse    [ 10752|  1500|  3584|   1926182|    0.181975|     29.5737|    0.251635
sparse    [ 10752|  3000|  3584|   1927537|    0.319327|     33.7302|    0.258285
sparse    [ 10752|  6000|  3584|   1925716|    0.648246|     33.1997|    0.226093
sparse    [  7680|     1|  2560|   1966992| 0.000147336|      24.867|    0.369858 
... 

```

Looking forward to your comments @v0dro, @pearu. 






[ghstack-poisoned]

aocsa commented on aten/src/ATen/native/mkl/SparseCsrLinearAlgebra.cpp:
```diff
@@ -105,21 +105,21 @@ class SparseCsrMKLInterface {
         dense,
-        dense_ncols,
-        dense_ncols,
+        N,
+        K,
```
I think it is not. This implementation is supposed to be right for matvec.

aocsa commented:
> Can you please make this change independent from the rest of the stack?
> And also, keep it restricted to PR title (i.e. just fixing addmm bug)

Thanks @malfet,  I am creating a new PR https://github.com/pytorch/pytorch/pull/58768 with the hope that it would be landed sooner.  I will close thise PR as well.
